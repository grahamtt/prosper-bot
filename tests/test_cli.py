import os
import sys
from decimal import Decimal

import pytest
from prosper_shared.omni_config import _realize_config_schemata, _realize_input_schemata
from prosper_shared.omni_config._define import _arg_parse_from_schema
from prosper_shared.omni_config._merge import _merge_config
from prosper_shared.omni_config._parse import _ArgParseSource

from prosper_bot.allocation_strategy import AllocationStrategies
from prosper_bot.bot import bot
from prosper_bot.cli import build_config


class TestCli:
    # Ensure the schema is loaded
    bot._schema()

    def test_build_config(self, mocker):
        mocker.patch.object(
            sys,
            "argv",
            [
                "test-cli",
                "--client-id=0123456789abcdef0123456789abcdef",
                "--username=fake-username",
                "--token-cache=fake-token-cache",
                "--dry-run",
                "--verbose",
                "--single-run",
                "--min-bid=30",
                "--target-loan-count=600",
                "--search-for-almost-funded",
                "--analytics",
                # TODO: the --strategy param is broken :(
                # "--strategy=CONSERVATIVE",
            ],
        )

        config = build_config()

        assert config._config_dict["prosper-bot"]["bot"] == pytest.approx(
            {
                "min-bid": Decimal("30"),
                "strategy": AllocationStrategies.AGGRESSIVE,
                "target-loan-count": 600,
                "search-for-almost-funded": True,
                "analytics": True,
            }
        )
        assert config._config_dict["prosper-bot"]["cli"] == {
            "dry-run": True,
            "verbose": True,
            "single-run": True,
        }
        assert config._config_dict["prosper-api"] == {
            "auth": {"token-cache": "fake-token-cache"},
            "credentials": {
                "client-id": "0123456789abcdef0123456789abcdef",
                "username": "fake-username",
            },
        }

    @pytest.mark.xfail("CI" in os.environ, reason="The output has changed on GitHub")
    def test_cli_help(self, mocker, snapshot):
        """This test asserts that the CLI help hasn't changed so we can ensure there are no backwards-incompatible changes."""
        mocker.patch(
            "prosper_shared.omni_config._define.user_config_dir",
            lambda i: f"/config_dir/dir/{i}",
        )
        mocker.patch("prosper_shared.omni_config._define.getcwd", lambda: "/cwd/dir")
        config_schemata = _merge_config(_realize_config_schemata())
        input_schemata = _merge_config(_realize_input_schemata())
        next(
            v
            for v in config_schemata["prosper-api"]["auth"].keys()
            if v._expected_val == "token-cache"
        )._default = "/some/path/to/token-cache"
        source = _ArgParseSource(
            _arg_parse_from_schema(
                config_schemata, input_schemata, prog_name="prosper-bot"
            )
        )
        assert source._argument_parser.format_help() == snapshot
