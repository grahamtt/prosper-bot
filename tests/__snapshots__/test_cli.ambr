# serializer version: 1
# name: TestCli.test_cli_help
  '''
  usage: prosper-bot [-h] [-c CLIENT-ID] [--client-secret CLIENT-SECRET]
                     [-u USERNAME] [-p PASSWORD] [-t TOKEN-CACHE] [-m MIN-BID]
                     [-s {AGGRESSIVE,CONSERVATIVE,OVERALL_HIGHEST_RATE}]
                     [--target-loan-count TARGET-LOAN-COUNT]
                     [--search-for-almost-funded] [-v] [-d] [--single-run]
  
  All optional program arguments can be provided via configuration file at the
  following locations: '/config_dir/dir/prosper-
  bot/config.{json|yml|yaml|toml}','/cwd/dir/prosper-
  bot.{json|yml|yaml|toml}',/cwd/dir/.pyproject.toml.
  
  optional arguments:
    -h, --help            show this help message and exit
  
  prosper-api.credentials:
    -c CLIENT-ID, --client-id CLIENT-ID
                          The client-id from Prosper; Type: str matching
                          /^[a-f0-9]{32}$/
    --client-secret CLIENT-SECRET
                          The client-secret from Prosper; can be stored and
                          accessed securely using the keyring library; Type: str
                          matching /^[a-f0-9]{32}$/
    -u USERNAME, --username USERNAME
                          Your Prosper username; Type: str
    -p PASSWORD, --password PASSWORD
                          Your Prosper password; can be stored and accessed
                          securely using the keyring library; Type: str
  
  prosper-api.auth:
    -t TOKEN-CACHE, --token-cache TOKEN-CACHE
                          The filesystem location where the auth token will be
                          cached; Type: str; Default: /some/path/to/token-cache
  
  prosper-bot.bot:
    -m MIN-BID, --min-bid MIN-BID
                          Minimum amount of a loan to purchase; Type: Decimal;
                          Default: 25.00
    -s {AGGRESSIVE,CONSERVATIVE,OVERALL_HIGHEST_RATE}, --strategy {AGGRESSIVE,CONSERVATIVE,OVERALL_HIGHEST_RATE}
                          Strategy for balancing your portfolio; Type: str;
                          Default: AGGRESSIVE
    --target-loan-count TARGET-LOAN-COUNT
                          Calculate min bid based on (total account value /
                          target loan count). Overrides min-bid if present;
                          Type: int
    --search-for-almost-funded
                          Search for listings with remaining funding <= cash,
                          which allows bidding when cash is less than $25; Type:
                          bool
  
  prosper-bot.cli:
    -v, --verbose         Prints additional debug messages; Type: bool
    -d, --dry-run         Run the loop but don't actually place any orders;
                          Type: bool
    --single-run          Runs the loop only once, or until cash is exhausted;
                          Type: bool
  
  '''
# ---
